[Sharing 2.0] create share

rullzer opened the pull request:
Creating a share with the new sharing code.

It sets the new uid_fileonwer column right away.

TODO:
- [X] Finish unit tests for more coverage
- [X] Inspect what might break
- ~~New integration tests?~~
- [x] HOOKS
  - [x] Preshare
  - [x] postshare
  - [x] password
- ~~Generate new target (do on initmount)~~
- [x] Target should have default share directory

rullzer commented:
So we need to properly generate targets. The question is where do we want to do this?

Currently there is this whole bounching around. From share/share.php => share/helper.php => files_sharing... => files_sharing/lib/helper.php etc

I think it would be best to handle this in the share provider. Since there it can be decided what to do. Since for example for federated sharing and maybe custom share providers there might be no (or a different) need when generating targets.

@schiesbn what do you think?

schiesbn commented:
Once a share gets created I would simply add the path of the owner as target and check for name collisions once the share gets mounted for the recipient and adjust the target then. This should have a much better performance instead of creating for all recipients (which can be a lot if you think about group shares) a unique target during the share operation.

rullzer commented:
OK agreed :)

rullzer commented:
Seems the current tests are failing because we do not copy over the skeleton dir properly if there is a mountpoint for that file in place.

rullzer commented:
The current failing OCS test is caused by not creating explicit reshares of files anymore.

schiesbn commented:
> Once a share gets created I would simply add the path of the owner as target and check for name collisions once the share gets mounted for the recipient and adjust the target then.

Just to clarify this part to make sure that we don't miss something. We shouldn't set the full owner path as target by default but only the file/folder name of the shared file/folder concatenated with the default target for shares (default is the root folder but it can be changed in config.php)

rullzer commented:
> Just to clarify this part to make sure that we don't miss something. We shouldn't set the full owner path as target by default but only the file/folder name of the shared file/folder concatenated with the default target for shares (default is the root folder but it can be changed in config.php)

Yes, I now just take the name of the file/folder. Should add the default folder. But indeed not the whole path.

rullzer commented:
Target is now constructed with share_dir

DeepDiver1975 commented:
@rullzer please rebase to get latest changes regarding the ocs api test execution in - maybe we see something of value in the logs why this breaks .... THX

rullzer commented:
@DeepDiver1975 rebased.

The only failing test I see is the one where we check for reshares. This is disabled for now. And we should be able to renable that once we implement getShares in the new code. Since that will list all reshares based on who owns the file etc.

rullzer commented:
@DeepDiver1975 @schiesbn @nickvergessen @PVince81 I know more test coverage is needed here. But I'd appriciate if you can already look at this PR. Especially the manager part (that holds the most 'logic'). And see if I'm missing something critical.

DeepDiver1975 commented on apps/files_sharing/api/share20ocs.php:
```diff
@@ -216,4 +336,34 @@ protected function canAccessShare(IShare $share) {
+	 * @return \DateTime
+	 */
+	private function parseDate($expireDate) {
+		if (preg_match('/^\d{4}-\d{2}-\d{2}$/', $expireDate) === 0) {
```
not really necessary

schiesbn replied:
yes, try-catch block below should already catch broken date strings.

rullzer replied:
Agreed removed

DeepDiver1975 commented on build/integration/features/sharing-v1.feature:
```diff
@@ -263,7 +265,7 @@ Feature: sharing
       | file_target | /textfile0.txt |
       | path | /textfile0.txt |
-      | permissions | 23 |
+      | permissions | 19 |
```
why?

rullzer replied:
Because of https://github.com/owncloud/core/issues/20839

Basically files do not have create permssion. And the new sharing code checks if you do not try to increase the permissions.

The old code happily set create permissions on files. Which was wrong.

DeepDiver1975 commented on build/integration/features/sharing-v1.feature:
```diff
@@ -326,7 +328,7 @@ Feature: sharing
     And As an "user1"
     When creating a share with
-      | path | /textfile0. (2).txt |
+      | path | /textfile0 (2).txt |
```
why?

rullzer replied:
There were two code paths to update the target. In the old code we sometimes generated the target at share time and sometimes only at init mount time.

Basically we need to do this at init mount time anyway. So there is no need to do it at share time as well. Also it does not scale to do it at share time (group shares for example). While at init mount time is fine.

nickvergessen replied:
> Basically we need to do this at init mount time anyway. So there is no need to do it at share time as well. Also it does not scale to do it at share time (group shares for example). While at init mount time is fine.

But that means any activities triggered before the first mount by the user will have invalid paths for them? :(

rullzer replied:
Yes, But that currently already happens for group shares anyway.

DeepDiver1975 commented on lib/private/share20/defaultshareprovider.php:
```diff
@@ -64,11 +64,104 @@ public function __construct(
+
+		if ($share->getShareType() === \OCP\Share::SHARE_TYPE_USER) {
+			//Set the UID of the user we share with
+			$qb->setValue('share_with', $qb->createParameter('shareWith'))
```
use createNamedParameter() instead of createParameter() followed by setParameter()

rullzer replied:
much better yes. Thanks.

Fixed

DeepDiver1975 commented on tests/lib/share20/managertest.php:
```diff
@@ -365,4 +383,362 @@ public function testGetShareById() {
+		try {
+			$this->manager->createShare($share);
+			$this->fail();
+		} catch (\InvalidArgumentException $e) {
```
I prefer to see here the use of https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.expectedException

nickvergessen replied:
You can also use `$this->setExpectedException()` to set it at a point late in the method, to make sure the pervious function calls don't throw

rullzer replied:
Ah, the reason I do it like this (in several places) is that I can then test if some stuff actually happen (including the exception). In this case even if the exception is thrown I want to know that the manager resets the time. If I do not wrap it in a try catch block the execution does not continue.

nickvergessen replied:
sounds valid

rullzer commented:
Ok let me finish the testing part so we can start to get this in.

rullzer commented:
Ok lets get this in so work on sharing can continue.

Calling all hands to review :)

schiesbn commented:
beside the preg_match() we can probably skip (see comment above) I think it is OK :+1:

rullzer commented:
Fixed comment. and rebased.

nickvergessen commented:
At the moment this breaks 4 of our smashbox tests:
- shareFile
- shareDir
- reshareDir
- shareGroup

Now the problem with that is, that if we merge it like this, we will be unable to spot further failures in other PRs, until that got fixed. The problem is after sharing the client syncs and verifies that the files exist, which they don't for reshares it seems.

sharePermission test passes, so that seems to work as intended.

nickvergessen commented:
Apart from that is looks good, activities are still triggered for group, user and link shares

nickvergessen commented on build/integration/features/sharing-v1.feature:
```diff
@@ -231,20 +231,22 @@ Feature: sharing
-    And User "user1" should be included in the response
-    And User "user2" should be included in the response
-    And User "user3" should not be included in the response
+# Skip this test for now. Since the new shares do not create reshares
```
Please add the magic word `TODO` or `FIXME` which IDEs highlight

rullzer commented:
What I could do is base getShares on this branch and finish that.. So we know that if we merge 2 PR's we only have a very small window of breakage

rullzer commented:
Ok I think i can fix some of the resharing. Which in turn would require less fixing the old code paths. It will require some fixing. But better to do it now.

rullzer commented:
Basically it goes wrong because the in app/files_sharing/lib/.. the old code is still in use. So they try to construct a path that is invalid.

The new resahres are created differently. Which is biting me now.

rullzer commented:
@nickvergessen could you smash again? Hopefully less fails this time...

rullzer commented:
Ok so it seems we need to also fix the smashbox tests.

Since we create flat reshares. So the behaviour is slightly different (so the tests should fail now).

@schiesbn can you have a look at the last commit?

rullzer committed:
[Share 2.0] Update db structure to stay backwards compatible

rullzer committed:
[Share 2.0] Update share class

More getters and setters are required to properly create shares

rullzer committed:
[Share 2.0] Make share provider ready for create shares

rullzer committed:
[Share 2.0] Make the share manager ready for share creation

rullzer committed:
[Share 2.0] Enable share creation via OCS API

rullzer committed:
Make sure to login at least once for the intergration tests

Else we run into race conditions with the skeleton code

rullzer committed:
Skip one intergration test until we have fixed the getshares

The new sharing code handles things differently so there is no way for
create shares to handle this all properly.

rullzer commented:
Ok, rebased, squased and all.. Lets see what jenkins thinks. I'll fix the getshares now (based on this branch). To make sure that the test breaking is as short as possible.

rullzer commented:
whoohoo all is happy (well except smashbox).

@schiesbn @nickvergessen @DeepDiver1975 please have a look. I did my best to collect all the checks spread over the place. I might miss some but I think this is starting to look complete.

Please review.

nickvergessen commented:
Working on smashbox atm I managed to fix
* shareFile - reshared file should exist after the shared file was unshared
* shareDir - reshared dir should exist after the shared file was unshared

However, then I applied the same logic change to:
* shareGroup - reshared dir should exist after the shared file was unshared

But that fails. are group shares untouched by your PR?

rullzer commented:
@nickvergessen that should work. Is your code somewhere so I can look at it?

nickvergessen commented:
https://github.com/owncloud/smashbox/pull/140

The failing tests case is due to a change in the error message:

* 8.2
> Status code: 403
> Message: You are not allowed to share /TEST_FILE_USER_RESHARE.dat/

* 9.0
> Status code: 404
> Message: Path is not shareable

I think the 404 is a mistake and it should in fact be 403, because the resource exists, its just forbidden to do it...

I can easily fix the smashbox assertion, but we need to be sure that the API change is intended

cc @schiesbn @rullzer @DeepDiver1975

nickvergessen commented on apps/files_sharing/api/share20ocs.php:
```diff
@@ -192,6 +191,127 @@ public function deleteShare($id) {
+		if ($shareType === \OCP\Share::SHARE_TYPE_USER) {
+			// Valid user is required to share
+			if ($shareWith === null || !$this->userManager->userExists($shareWith)) {
+				return new \OC_OCS_Result(null, 404, 'please specify a valid user');
```
please check, that all the status codes still matches the old status codes

nickvergessen commented on apps/files_sharing/api/share20ocs.php:
```diff
@@ -192,6 +191,127 @@ public function deleteShare($id) {
+		try {
+			$share = $this->shareManager->createShare($share);
+		} catch (\Exception $e) {
+			return new \OC_OCS_Result(null, 404, $e->getMessage());
```
this is where we get the 404 instead of the old 403

rullzer commented:
@nickvergessen OCS status codes !== HTTP status codes

https://doc.owncloud.org/server/8.2/developer_manual/core/ocs-share-api.html

400 - unknown share type
403 - public upload was disabled by the admin
404 - file couldn’t be shared

so I can change it. (should not be to much work). But the original 403 is wrong basically. (according to the docs).

I'm fine with fixing either. But this is exactly the reason why we need proper HTTP status codes...

rullzer commented:
Anyway, lets make sure the implementation does not change for now. So I'll fix the 404 => 403

jospoortvliet commented:
:rebase: -> why is that not a emoticon here?!?!? Tsssk.

rullzer committed:
[Sharing 2.0] General exceptions return 403

This is the same as the old behaviour

rullzer commented:
Ok, so apparently we threw a 403 by default. This is fixed now... now trying to get all places we throw something else...

rullzer committed:
[Sharing 2.0] Do not use static function to get numeric storage id

rullzer committed:
[Sharing 2.0] Add L10N instance to manager for translated errors

rullzer commented:
Ok, I *think* smashbox should be happier now.
@nickvergessen if you have time :)

nickvergessen commented:
Smashed it, and it passed on https://github.com/owncloud/smashbox/pull/140 now.

rullzer committed:
[Sharing 2.0] Some error cases report 404 instead of 403

jospoortvliet commented:
\o/ :joy:

rullzer commented:
Now that smashbox is passing as well. Please review once more. @schiesbn @nickvergessen @DeepDiver1975

schiesbn commented:
all tests passed and code looks good at first sight, so let's get this in to move forward :+1:

DeepDiver1975 commented:
> I can easily fix the smashbox assertion, but we need to be sure that the API change is intended

no api change!

DeepDiver1975 commented:
:+1:
